Gateway & Dashboard 5.11.0 RNs.mdx (#1018)
TykTechnologies/tyk-docs: developer-support/release-notes/gateway.mdx

Our minor releases are supported until our next minor comes out.

---

## 5.11 Release Notes 

### 5.11.0 Release Notes 

#### Release Date xx December 2025

#### Release Highlights

> 

Reviewer: ```suggestion

Tyk 5.11 delivers security enhancements and deeper operational visibility - empowering teams to scale their API programs with confidence and efficiency.

**Strengthened API Security & Authentication**
This release advances our security foundation with enhanced JWT authentication capabilities. Teams can now leverage scope-to-policy mapping without requiring default policies, while new support for nested claims enables more granular policy and subject identification. We've also added IP spoofing protection through configurable depth selection in X-Forwarded-For headers.

**Advanced Operational Control**
Operations teams gain greater flexibility with the ability to temporarily remove targets from upstream load balancers during maintenance windows. Enhanced observability comes through OTel trace propagation to custom gRPC plugins, trace ID inclusion in API traffic logs, and dedicated Gateway latency metrics alongside upstream measurements. Data Plane Gateways now recover more quickly from interruptions to the MDCB link to the Control Plane.

**Enhanced Stability & Performance**
This release includes important stability improvements, resolving crash conditions in JWT authentication and concurrent processing scenarios, eliminating blocking operations that could cause significant delays during MDCB connectivity issues, and improving performance for OAuth key retrieval in hybrid deployments. These fixes collectively deliver a more reliable and responsive API gateway experience for enterprise environments.

These enhancements collectively strengthen Tyk's position as the platform of choice for organizations requiring enterprise-scale API management with robust security, operational excellence, and developer productivity.
```

</Accordion>

<Accordion title='Add Trace ID to Traffic Logs for Improved Observability'>
When OpenTelemetry is enabled, the Trace ID allocated to each request is tagged in traffic logs as `trace-id-{traceID}` and also exposed in `$tyk_context.tyk_trace_id`. This makes it easier to correlate traffic logs with OTel traces in observability platforms and also allows transformation middleware and custom plugins to correlate logs with traces.
</Accordion>

<Accordion title='Enhanced `X-Forwarded-For` IP Selection with Configurable Depth Control'>
Added `xff_depth` configuration parameter to the Gateway's HTTP server options, enabling operators to specify which IP address from the `X-Forwarded-For` header chain should be considered the real client IP. The feature counts from the rightmost (most trusted) end of the IP chain for improved security while maintaining backward compatibility. When `xff_depth` is set to 0 or not configured, Tyk continues using the first IP address as before. Updated default configurations across Tyk Demo, Helm Charts, and example files to use `xff_depth=1` for enhanced security in new deployments.
</Accordion>

> <Accordion title='Added JWKS cache flush to the Dashboard API and MDCB'>

Reviewer: ```suggestion
```

> Added JWKS cache invalidation endpoint `DELETE /api/cache/jwks/{apiID}` to the Dashboard API to flush JWKS cache for specific APIs across all connected Gateways. This extends the existing Gateway-level JWKS cache flush functionality to work through the Dashboard API with proper MDCB propagation to data plane Gateways.

Reviewer: ```suggestion
```

> </Accordion>

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

#### Downloads

- [Docker image to pull](https://hub.docker.com/r/tykio/tyk-gateway/tags?page=&page_size=&ordering=&name=v5.11.0)
  - ```bash
    docker pull tykio/tyk-gateway:v5.11.0
    ``` 
- Helm charts
  - [tyk-charts v4.2.0](/developer-support/release-notes/helm-chart#420-release-notes)

Please note that the Tyk Helm Charts are configured to install the LTS version of Tyk Gateway. You will need to modify them to install v5.11.0.

- [Source code tarball of Tyk Gateway v5.10.1](https://github.com/TykTechnologies/tyk/releases/tag/v5.11.0)

#### Changelog
<a id="Changelog-v5.11.0"></a>

##### Added

<AccordionGroup>

<Accordion title='Support for Nested JWT Claim Paths in Subject, Policy, and Scope Mapping'>
Added support for nested JWT claims for subject and policy fields, enabling hierarchical claim structures to be used in authentication and policy application. Now you can configure the `subjectClaims`, `basePolicyClaims`, and `scopes.claims fields` to use nested claim names, such as `test.sub` or `policy.base`.
</Accordion>

<Accordion title='Enhanced Latency Metrics with Gateway-Only Processing Time'>

> - Enhanced request-level timing by tracking precise timestamps when requests enter the Gateway, enabling accurate end-to-end latency calculations that extend beyond previous proxy-only measurements.

Reviewer: ```suggestion
We have enhanced request-level timing by tracking precise timestamps when requests enter the Gateway, enabling accurate end-to-end latency calculations that extend beyond previous proxy-only measurements.
```

</Accordion>

<Accordion title='Support Temporary Removal of Upstream Targets via Zero-Weight Load Balancing'>
You can now temporarily remove upstream targets (servers) from Tyk's upstream load balancing group. If a target is removed from the group, Tyk will route no traffic to it. This allows temporary target removal for maintenance, troubleshooting or environment issues.
Simply set the weight for the target to zero, and it will be removed from the round robin list. Multiple targets can be removed, but at least one must have a non-zero weight and thus will be served traffic.
</Accordion>

<Accordion title='Removed Default Policy Requirement for JWT Scope-to-Policy Mapping'>
We’ve removed the need to supply a default policy when using scope-to-policy mapping with JWT Authentication. Now, if you enable scope-to-policy mapping by configuring `scopes.claimName`, you do not need to provide a policy ID in `defaultPolicies`. If a request does not contain any valid scopes, it will be rejected with `HTTP 403 Forbidden` (default deny). You can still provide a default policy if you require a different behaviour.
</Accordion>

> <Accordion title='Add Trace ID to Traffic Logs for Improved Observability'>

Reviewer: ```suggestion
<Accordion title='Add OTel Trace ID to Traffic Logs for Improved Observability'>
```

When OpenTelemetry is enabled, the Trace ID allocated to each request is tagged in traffic logs as `trace-id-{traceID}` and also exposed in `$tyk_context.tyk_trace_id`. This makes it easier to correlate traffic logs with OTel traces in observability platforms and also allows transformation middleware and custom plugins to correlate logs with traces.
</Accordion>

> <Accordion title='Enhanced `X-Forwarded-For` IP Selection with Configurable Depth Control'>

Reviewer: ```suggestion
<Accordion title='Added Configurable `X-Forwarded-For` Header Selection'>
```

> Added `xff_depth` configuration parameter to the Gateway's HTTP server options, enabling operators to specify which IP address from the `X-Forwarded-For` header chain should be considered the real client IP. The feature counts from the rightmost (most trusted) end of the IP chain for improved security while maintaining backward compatibility. When `xff_depth` is set to 0 or not configured, Tyk continues using the first IP address as before. Updated default configurations across Tyk Demo, Helm Charts, and example files to use `xff_depth=1` for enhanced security in new deployments.

Reviewer: ```suggestion
Added `xff_depth` configuration parameter to the Gateway's HTTP server options for improved security. This specifies which entry in the `X-Forwarded-For` header chain should be considered to contain the real client IP. The value set in `xff_depth` is used when counting from the rightmost (most trusted) end of the IP chain, where a value of 1 is the first (counting from the right). If `xff_depth` is set to 0 or not configured, Tyk continues using the first IP address as before, maintaining backward compatibility. We have updated the default configurations across Tyk Demo, Helm Charts, and example files to use `xff_depth=1` for enhanced security in new deployments.
```

</Accordion>

> <Accordion title='Added JWKS cache flush to the Dashboard API and MDCB'>

Reviewer: Not a Gateway change (if there's a minor tweak in Gateway, that's just to support the Dashboard feature, doesn't need to be logged here).

Author: Ticket had Tyk Gateway as a component, I'll remove it
